{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### IMPERIAL COLLEGE LONDON\n",
    "\t\n",
    "## Timed Assessment for the Master of Science in\n",
    "## Bioinformatics and Theoretical Systems Biology 2021-22 \n",
    "\n",
    "---\n",
    "\n",
    "## Python Timed Assessment\n",
    "\n",
    "\n",
    "__This assessment provides 60% of the marks for the Computing section.__\n",
    "\n",
    "__Please answer all three questions.__\n",
    "\n",
    "__Tuesday 26th April 2022 (10:00 – 13:00)__\n",
    "\n",
    "__At the end of the assessment email your complete Jupyter Notebook file to d.huntley@imperial.ac.uk__\n",
    "\n",
    "--- \n",
    "\n",
    "__You are required to comment the code in each of your answers as this will make them more readable and easier to mark.__\n",
    "\n",
    "__You should not over engineer the code.__\n",
    "\n",
    "__You may use any notes, books or the Internet to complete the exam__\n",
    "## Note that you cannot use any modules that are not included by default with Anaconda Python  \n",
    "\n",
    "---\n",
    "\n",
    "## Question 1\n",
    "\n",
    "The code below is required to predict all the open reading frames (ORF) in the forward strand only of the __bacteria_seq.fasta__ sequence. \n",
    "\n",
    "The requirements for an ORF are:\n",
    "\n",
    "A start codon of ATG<br>\n",
    "A stop codon of TAA, TAG or TGA<br>\n",
    "The ORF is at least 30 codons in length.\n",
    "\n",
    "The script currently fails to run as it contains numerous errors. Fix the errors in the script so that it will predict the ORFs in the __bacteria_seq.fasta__ sequence.\n",
    "\n",
    "The script contains both syntax errors, that prevent it from running, and logic errors, that cause it to incorrectly predict the ORFs.\n",
    "\n",
    "You should comment each modification you make to the script.\n",
    "\n",
    "You should only correct the errors in the code and not include any other modifications.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Q1.\n",
    "#my own code: testing. \n",
    "with open('bacteria_seq.fasta') as in_file: #changed the name. \n",
    "    seq = \"\"\n",
    "    for line in in_file:\n",
    "        line = line.rstrip() #added this\n",
    "        #print(\"lineee\", line)\n",
    "        \n",
    "        if not line.startswith('>'):\n",
    "            seq += line\n",
    "    #print(\"seq\", seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 1\n",
    "\n",
    "with open('bacteria_seq.fasta') as in_file:  #file name was incorrect. changed from infile -> in_file. \n",
    "    seq = \"\"  #added seq empty string here.\n",
    "    for line in in_file:\n",
    "        line = line.rstrip()  #added this, remove new line charachters. \n",
    "        #print(\"line:\", line)  #added this line. \n",
    "        if not line.startswith('>'):\n",
    "            seq += line\n",
    "    #print(seq) #added this. \n",
    "\n",
    "for i in range(0, 3): #changed the range from 1->3, into 0 -> 3. \n",
    "    start = 0\n",
    "    for j in range(i, len(seq) - 2, 3): #added the step '3'.\n",
    "        codon = seq[j:j+3]\n",
    "        #print(\"codon:\", codon)   #added this print function.\n",
    "\n",
    "        if codon == 'ATG' and start == 0:  #changing the indending of the IF statement.\n",
    "            start = j\n",
    "    \n",
    "        elif codon == 'TAA' or codon == 'TAG' or codon == 'TGA' and start != 0:  #changed else to elif.\n",
    "        \n",
    "            if j - start > 30:\n",
    "                \n",
    "                print(\"Reading frame:\", i+1, \"starts at:\", start+1, \"end at\", j+1)   #added the brackets around the print function. added 'Reading frame', 'Starts at', and 'ends at' as well. \n",
    "\n",
    "        start = 0  \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Question 2\n",
    "\n",
    "The __test.matrix__ file contains a nucleotide scoring matrix and the test.align file contains a simple sequence alignment. \n",
    "\n",
    "Complete the code below so that it calculates the score for the sequence alignment in __test.align__ using the scoring matrix in __test.matrix__.  \n",
    "\n",
    "Your script should be able to work with any scoring matrix in the same format so is required to read the __test.matrix__ file.\n",
    "\n",
    "The penalty to open a gap is -5 and for each extension is -2.\n",
    "\n",
    "When the score has been calculated the script should then calculate the associated E-value. The equation to calculate the E-value is:\n",
    " \n",
    "![title](Evalue.jpg)\n",
    " \n",
    "K and lambda are matrix specific values. S is the score, m the length of the query sequence and n the length of the subject sequence. All of these values are defined in code.\n",
    "\n",
    "The score and E-value should be printed to screen.\n",
    "\n",
    "The script should concisely address the requirements of the question and should not be substantially altered to add additional features.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 2\n",
    "\n",
    "import math\n",
    "import numpy as np #module in anaconda python.\n",
    "import pandas as pd #python data analysis library in anaconda python\n",
    "\n",
    "#lambda = 0.37\n",
    "K = 0.127\n",
    "m = 110\n",
    "n = 90\n",
    "#other way:\n",
    "#d = {\"A\":0, \"T\":1, \"C\":2, \"G\":3}  could make a dictionary?\n",
    "\n",
    "#reading in the file:\n",
    "with open('test.matrix') as in_file:\n",
    "\n",
    "    line = in_file.readlines()  #read in the file. \n",
    "    line =line[1:]              #removes the first line as its not necessary.\n",
    "    for line in line:\n",
    "        line = line.split()     #split the lines. \n",
    "        \n",
    "        line = np.array(line)   #A numpy array is a grid of values.\n",
    "    print(\"matrix:\", line)             #printing the matrix. \n",
    "\n",
    "#The score of an alignment is calculated as the sum of substitution and gap scores.\n",
    "#The test.matrix file is a triangular matrix. \n",
    "\n",
    "with open('test.align') as in_file2:  #reading in the 2nd file, which is the alignment file. \n",
    "    \n",
    "    #Was initially trying to use numpy and pandas, but got a bit confused and spent too much time reading the documentation.\n",
    "    \n",
    "    #matrix_a = np.loadtxt(line)\n",
    "    #matrix_a= pd.DataFrame(aligned_matrix) \n",
    "\n",
    "\n",
    "    line2 = in_file2.readlines() #read in the lines.\n",
    "    for line in line2:           #work through the list.\n",
    "        sequence1 = line[0].split()  #split the elements the are present in the first sequence\n",
    "        sequence1 = sequence1[2]     #get element number 2 in sequence1, and have that as the sequence. \n",
    "        sequence2 = line[2].split()  #split the elements.\n",
    "        sequence2 = sequence2[2]\n",
    "        sequence_both = list(zip(sequence1,sequence2))   #returns a zip object. We can then iterate through tuples, where all the values that have been passed as arguments are stored as pairs\n",
    "        count = 0\n",
    "    \n",
    "    for i in range(0, len(sequence_both)):  #go through the list.\n",
    "        if sequence_both[i][0] == \"A\" and sequence_both[i][0] == \"T\":\n",
    "            score = 1\n",
    "        \n",
    "    #this is where I could get to in the time allocated. \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Question 3\n",
    "\n",
    "The EMBOSS program pepinfo provides information about amino acid properties, such as hydropathy or charged residues, and their position in a protein sequence. This program has been run on a sequence and the results for positively charged amino acids and polar amino acids are in the __pepinfo.positive__ and __pepinfo.polar__ files respectively. The columns in the file are the position in the sequence, the amino acid and the result – 1 if it has the property and 0 otherwise.\n",
    "\n",
    "Pepinfo also provides a plot of all the amino acid types found and their location. An example of these plots is shown below:\n",
    "\n",
    "![title](pepinfo.jpg)\n",
    "\n",
    "Complete the code below so that it reads the __pepinfo.positive__ and __pepinfo.polar__ files and produces a plot to show all the amino acids which are both positive and polar. \n",
    "\n",
    "The script should concisely address the requirements of the question and should not be substantially altered to add additional features.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 3\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#reading in the first file:\n",
    "with open('pepinfo.positive') as in_file1:  #reading in the 2nd file, which is the alignment file. \n",
    "\n",
    "    line1 = in_file1.readlines()[1:]          #read in the file.\n",
    "    line1 = [i.rstrip() for i in line1]\n",
    "    #split lines. \n",
    "    position = [int(i.split()[0]) for i in line1]  #getting the position.\n",
    "    \n",
    "#reding in the second file:\n",
    "\n",
    "with open('pepinfo.polar') as in_file2:  #reading in the 2nd file, which is the alignment file. \n",
    "\n",
    "    line2 = in_file2.readlines()[1:]\n",
    "    line2 = [i.rstrip() for i in line2]  \n",
    "    polar = [int(i.split()[0]) for i in line2]    \n",
    "    \n",
    "    \n",
    "    \n",
    "#we need to match the two files:\n",
    "aa_pos_polar = []  #initialise a list of amino acids that are both positive and polar.\n",
    "\n",
    "\n",
    "for i,j in zip(position,polar):   #could not get zip to work in Q2, but will utilise it here. \n",
    "    if i == 1 and j == 1:     # '1' if property is present.\n",
    "        aa_pos_polar.append(1) #append to the list. \n",
    "    else:\n",
    "        aa_pos_polar.append(0)\n",
    "\n",
    "\n",
    "#plotting the graph:\n",
    "\n",
    "\n",
    "plt.bar(position, aa_pos_polar) \n",
    "plt.xlabel(\"aa pos and polar residues\")\n",
    "plt.title(\"aa pos and polar residues\")\n",
    "plt.show()\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Email your completed notebook file to d.huntley@imperial.ac.uk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
